File: ext/nim/examples/defer.nim
## ext/nim/examples/defer.nim —— 与 go / sbcl / vlang 那三份 defer 例子**同一件事**
##
## 期望输出逐行相同：in / b / a / out。
## 到这儿 scope-exit 那一格已经有**四门语言、四种语法**落在同一格节点上了 ——
## G5 那条"提供者名单 ≥ 4 才算一台机器"（ADR-0033 §3.7）在这一格上成立。

proc demo() =
  defer: echo "a"
  defer: echo "b"
  echo "in"
  return

demo()
echo "out"

File: ext/go/examples/defer.go
// ext/go/examples/defer.go —— **scope-exit** 那一格的例子（第三个例子家族）
//
// 期望输出（与 ext/sbcl/examples/defer.lisp 逐行相同）：
//   in
//   b
//   a
//   out
//
// 三条语义各压一格，而且三条都由调度器给、不由语言给：
//   注册那一刻记下动作 · 宿主 region 出口时**逆序**跑 · **早退（return）也跑**

package main

import "fmt"

func demo() {
	defer fmt.Println("a")
	defer fmt.Println("b")
	fmt.Println("in")
	return
}

func main() {
	demo()
	fmt.Println("out")
}
